multispecies_whale_detection/dataset.py

```diff
@@ -70,3 +92,257 @@ def new(tfrecord_filepattern: str):
+    context_duration_seconds: float,
+    sample_size: int,
+    class_names: Sequence[str],
+    min_overlap=DEFAULT_MIN_OVERLAP) -> Tuple[tf.Tensor, tf.Tensor]:
+  """Extracts labeled context windows with random starts.
+
+  Args:
+    features: Features of an audio clip as returned by :py:meth:parse_fn.
+    context_duration_seconds: Desired duration of extracted windows. All audio
+      tensors returned will have exactly this duration.
+    sample_size: Desired number of windows to extract.
+    class_names: A list of all possible values of ANNOTATION_LABEL, also known
+      as the label vocabulary.
+    min_overlap: Lower bound for duration of overlap between an annotation
+      interval and an extracted window for the window to be considered positive
+      for the class in that annotation's label.
+
+  Returns:
+    Audio tensor of shape [sample_size, audio_duration_samples] and labels
+    tensor of float32 class labels of shape [sample_size, len(class_names)].
+  """
+  extracted_windows = []
+  extracted_labels = []
+  begin_limit = _audio_duration(features) - context_duration_seconds
+  for _ in range(sample_size):
+    begin = tf.random.uniform([], tf.constant(0.0, tf.float32), begin_limit)
+    window, label = extract_window(features, begin, context_duration_seconds,
+                                   class_names, min_overlap)
+    extracted_windows.append(window)
+    extracted_labels.append(label)
+
+  return tf.stack(extracted_windows), tf.stack(extracted_labels)
+
+
+def sliding_windows(
+    features: Dict[str, tf.Tensor],
+    context_duration_seconds: float,
+    hop_seconds: float,
+    class_names: Sequence[str],
+    min_overlap=DEFAULT_MIN_OVERLAP) -> Tuple[tf.Tensor, tf.Tensor]:
+  """Extracts labeled context windows sliding at a fixed period.
+
+  The number of windows extracted depends on the duration of the audio given.
+  All windows that fit within the bounds of the clip will be returned.
+
+  Args:
+    features: Features of an audio clip as returned by :py:meth:parse_fn.
+    context_duration_seconds: Desired duration of extracted windows. All audio
+      tensors returned will have exactly this duration.
+    hop_seconds: The fixed period at which the slice to extract slides over the
+      audio clip.
+    class_names: A list of all possible values of ANNOTATION_LABEL, also known
+      as the label vocabulary.
+    min_overlap: Lower bound for duration of overlap between an annotation
+      interval and an extracted window for the window to be considered positive
+      for the class in that annotation's label.
+
+  Returns:
+    Audio tensor of shape [sample_size, audio_duration_samples] and labels
+    tensor of float32 class labels of shape [sample_size, len(class_names)].
+  """
+  num_hops = 1 + tf.cast(
+      tf.math.floor(
+          (_audio_duration(features) - context_duration_seconds) / hop_seconds),
+      tf.int32)
+  extracted_windows = tf.TensorArray(tf.float32, size=num_hops)
```

Comment:
(neat! I don't think I've seen these before.)

Reply:
Thanks.

It was mostly new to me as well, though I think I've seen it once before around a while loop in the PCEN IIR filter.

My first attempt here used ```for _ in range(num_hops)```, and it caused an error of the form "this is not in the same graph as that." I think this is because num_hops is not statically known because the clip duration isn't. There's an option to add that constraint and make users pass a static clip duration down from the main method (flags), but I don't think this is so much less efficient as to warrant that tiny extra burden.

(I haven't tested this on TPU and would need to be sure it works.)